pkg/client/unversioned/urlbackoff.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package unversioned
+
+import (
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/util"
+	"net/url"
+	"time"
+)
+
+// URLBackoff struct implements the semantics on top of Backoff which
+// we need for URL specific exponential backoff.
+type URLBackoff struct {
+	//Uses backoff as underlying implementation.
+	Backoff *util.Backoff
+}
+
+// Clear clears the backoff data for all Urls.  This is useful for unit tests
+// or other environments where we might want to start retrying at a fast rate.
+func (b *URLBackoff) Reset() {
+	b.Backoff = util.NewBackOff(1*time.Second, 64*time.Second)
+}
+
+// Disable makes the backoff trivial, i.e., sets it to zero.  This might be used
+// by tests which want to run 1000s of mock requests without slowing down.
+func (b *URLBackoff) Disable() {
+	glog.V(4).Infof("Disabling backoff strategy")
+	b.Backoff = util.NewBackOff(0*time.Second, 0*time.Second)
+}
+
+// Every url maps to a key, which is a subset of that url.
+// For example, 127.0.0.1:8080/api/v2/abcde -> 127.0.0.1:8080.
+func (b *URLBackoff) baseUrlKey(rawurl string) string {
+	//Simple implementation for now, just the host.
+	//We may backoff specific paths (i.e. "pods") differentially
+	//in the future.
+	host, err := url.Parse(rawurl)
+	if err != nil {
+		glog.V(4).Infof("Error extracting url: %v", rawurl)
+		return ""
+	}
+	return host.Host
+}
+
+// updateBackoff updates backoff metadata
+func (b *URLBackoff) UpdateBackoff(actualUrl string, err error, responseCode int) {
+	//range for retry counts that we store is [0,13]
+	if responseCode >= 300 || err != nil {
```

Comment:
@timothysc better, complete reference: http://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml

@jayunit100 I'm suggesting you avoid encoding retries for requests that cannot be retried at this level. A 400 error code is a bad request. Retrying the exact same request will almost always cause a 400. It is up to the client to fix their error, and then resend. I also think that this backoff should only encode a wait for errors that actually indicate the server is overloaded. With the exception of 429, none of the 300 or 400 codes indicate that.